torch/lib/THNN/generic/ClassNLLCriterion.c

```diff
@@ -115,20 +135,54 @@ void THNN_(ClassNLLCriterion_updateGradInput)(
     THError("weight tensor should be defined either for all or no classes");
   }
 
+  if (!reduce && n_dims == 2) {
+    int batch_size = THTensor_(size)(input, 0);
+    THNN_CHECK_DIM_SIZE(gradOutput, 1, 0, batch_size);
+
+    int i;
+    #pragma omp parallel for private(i)
+    for (i = 0; i < batch_size; i++) {
+      int cur_target = THIndexTensor_(get1d)(target, i);
+      if (cur_target == ignore_index) {
+        continue;
+      }
+      real weight = weights ? THTensor_(get1d)(weights, cur_target) : 1.0f;
+      THTensor_(set2d)(gradInput, i, cur_target, -weight * THTensor_(get1d)(gradOutput, i));
+    }
+    return;
+  }
+
+  if (!reduce && n_dims <= 1) {
+    sizeAverage = false;
+  }
+
+  real *total_weight_data = THTensor_(data)(total_weight);
+
+  if (!(*total_weight_data > 0)) {
```

Comment:
It seems that `total_weight_data <= 0` would be clearer. Also, what's the purpose of this check?

Reply:
Agreed. I'm not really sure what the purpose of this check is (it was in the [non-reduce code](https://github.com/pytorch/pytorch/blob/d5604aea0bf23c47a915f470c903beb3b6915090/torch/lib/THNN/generic/ClassNLLCriterion.c#L105) before. I'll investigate a little and see if it's necessary.